Design.md

```diff
@@ -0,0 +1,149 @@
+# Design doc for job-runner
+
+Job-runner provides a gRCP API to execute arbitrary commands on a Linux host.
+
+## Runner
+
+The jobs themselves are normal child processes of the runner. The runner configures CGroups and Linux namespaces for children to limit resource usage and isolate the them from the other processes. It also handles output recording and forwarding. All data is stored in-memory. In case of the service process termination, all jobs and their status are lost.
+
+## Resource limits and isolation
+
+The job-runner uses cgroups V1 for resource limits. Cgroups will be used through an API wrapper like [controlgroup-rs](https://github.com/ordovicia/controlgroup-rs) or [cgroups-fs](https://github.com/frol/cgroups-fs). There are many tweakable options for limiting resource use, but the following minimal set has been selected to demonstrate how they work. In particular, the following cgroup parameters will be set:
```

Comment:
> Cgroups will be used through an API wrapper

I think we usually ask to implement cgroups functionality without relying on 3rd party libraries but others may correct me if I'm wrong @knisbet @awly @russjones

Reply:
I'll roll my own then, although if I'm going to do that, I'll use the simpler v2 API.

Reply:
Updated